crates/m3/src/builder/indexed_lookup.rs

```diff
@@ -0,0 +1,431 @@
+// Copyright 2025 Irreducible Inc.
+
+use std::iter;
+
+use binius_core::constraint_system::channel::{Boundary, ChannelId, FlushDirection};
+use binius_field::{Field, PackedExtension, PackedField, TowerField};
+
+use super::{
+	constraint_system::ConstraintSystem, error::Error, witness::WitnessIndex, B1, B128, B16, B32,
+	B64, B8,
+};
+
+/// Indexed lookup tables are fixed-size tables where every entry is easily determined by its
+/// index.
+///
+/// Indexed lookup tables cover a large and useful class of tables, such as lookup tables for
+/// bitwise operations, addition of small integer values, multiplication, etc. The entry encodes
+/// an input and an output, where the index encodes the input. For example, a bitwise AND table
+/// would have 2 8-bit input values and one 8-bit output value. The index encodes the input by
+/// concatenating the 8-bit inputs into a 16-bit unsigned integer.
+///
+/// This trait helps to count the number of times a table, which is already filled, reads from a
+/// lookup table. See the documentation for [`tally`] for more information.
+pub trait IndexedLookup<F: TowerField> {
+	/// Binary logarithm of the number of table entries.
+	fn log_size(&self) -> usize;
+
+	/// Encode a table entry as a table index.
+	fn entry_to_index(&self, entry: &[F]) -> usize;
+}
+
+/// Determine the read counts of each entry in an indexed lookup table.
+///
+/// Before a lookup table witness can be filled, the number of times each entry is read must be
+/// known. Reads from indexed lookup tables are a special case where the counts are difficult to
+/// track during emulation, because the use of the lookup tables is an arithmetization detail. For
+/// example, emulation of the system model should not need to know whether integer additions within
+/// a table are constraint using zero constraints or a lookup table for the limbs. In most cases of
+/// practical interest, the lookup table is indexed.
+///
+/// The method to tally counts is to scan all tables in the constraint system and boundaries
+/// values, and identify those that pull from the lookup table's channel. Then we iterate over the
+/// values read from the table and count all the indices.
+///
+/// ## Returns
+///
+/// A vector of counts, whose length is equal to `1 << indexed_lookup.log_size()`.
+pub fn tally<P>(
+	cs: &ConstraintSystem<B128>,
+	// TODO: This doesn't actually need mutable access. But must of the WitnessIndex methods only
+	// allow mutable access.
+	witness: &mut WitnessIndex<P>,
+	boundaries: &[Boundary<B128>],
+	chan: ChannelId,
+	indexed_lookup: &impl IndexedLookup<B128>,
```

Comment:
Feels like could be `dyn`.

Reply:
Maybe. Is there a benefit to dyn > impl here?